### DISCUSSION TITLE
`execution_options` on an AsyncConnectable

### DISCUSSION THREAD

**Discussion by jellevwezel** [sqlalchemy/sqlalchemy#10059] [Usage Questions] [answered]
Hi,

The `execution_options` method for an `AsyncEngine` differs in return type from the `AsyncConnection`.
For the `AsyncConnection` an `async` method is used and for the `AsyncEngine` the method is `sync`:

```python
# AsyncConnection
@overload
async def execution_options(self, **opt: Any) -> AsyncConnection:
    ...

# AsyncEngine
@overload
def execution_options(self, **opt: Any) -> AsyncEngine:
    ...
```

In my code, I expect an `AsyncConnectable` and create a `sub bind` with the `execution_options` however when the code gets an `AsyncEngine` the behavior now differs from when it gets an `AsyncConnection`, while both are an `AsyncConnectable`.

The `AsyncConnectable` does not mention anything about the `execution_options` however I assumed they would have the same return type.

Would it be preferable/possible to have them both use either `sync` or `async`?

**Comment by jellevwezel** [comment]
With "return type" I meant a coroutine vs just the object.

**Comment by zzzeek** [comment]
`AsyncConnectable` is a common interface that doesn't really have any methods on it, it's mostly a marker at this time for some internal characteristics.

Overall I would not advise writing code that accepts `AsyncConnectable` - the pattern of "accepts an engine or connection" is a very old and very much legacy SQLAlchemy pattern which relied upon many methods that have since been removed in SQLAlchemy 2.0.    Code that works with engines vs. code that works with connections/transactions are very different in their role.

The engine's version of execution_options does not result in any SQL operations being invoked, and is more of a top-level configuational pattern such as in the example at https://docs.sqlalchemy.org/en/20/core/connections.html#maintaining-multiple-isolation-levels-for-a-single-engine ,  whereas the one on connections in some cases can.

**Reply by jellevwezel** [reply]
I understand your point, some context:

This question came from my way of unit testing, where I want to roll back my transaction after each test.
I achieve this by creating a connection from an engine and starting a 'root_transaction' then I create an `AsyncSession` with `join_transaction_mode='create_savepoint'` in order to have nested transactions.

When I create an `AsyncSession` context, in the test or somewhere in my code, I can now bind the connection to it and rollback after the test finishes. Where normally, I would bind the `AsyncSession` to an `AsyncEngine`. This is what I use the `AsyncConnectable` for (just for typing, because the `AsyncSession` needs an `AsyncEngine` or an `AsyncConnection` for its bind), and because my system uses multi schema it needs to change the `schema_translate_map` with the `execution_options`.

So when my code tries to bind the connectable to the `AsyncSession`, it complains that it cannot bind a coroutine during my unit tests since I simply bind the connectable to the `AsyncSession`, i.e.:

```python

connectable = engine or connection

session_factory = async_sessionmaker(
            autocommit=False,
            autoflush=False,
            expire_on_commit=False
            # join_transaction_mode='create_savepoint'  # during testing
        )

@asynccontextmanager
async def get_session(schema) -> AsyncIterator[AsyncSessionClass]:
   schema_translate_map = get_schema_translate_map(schema)
   sub_bind = connectable.execution_options(schema_translate_map=schema_translate_map)
   async with session_factory(bind=sub_bind) as session:
       yield session
```

Fixing this on my end is trivial, however I wondered why the `execution_options` differed for the two objects.
Please let me know if the same effect of using a root transaction can be used with only an engine.

**Reply by zzzeek** [reply, marked as answer]
if the API design here were to be "fixed", it would be that the two methods would be named completely differently, because they actually do different things.  `execution_options()` on an `Engine` is setting up the options that will be used *on the next connection, when it connects*.  Whereas `execution_options()` on a `Connection` is setting up the options *right now, in the current transaction*.   So the latter implies SQL commands whereas the former does not.  

so if anything, they should not be named the same.   something we can look into, figuring out two names that express these two different implications.

**Reply by Kokleng-Dev** [reply]
for me execution_options work only once after commit, it will go to default schema again.
but if i put it twice, it work fine. how to solve it.

```
  await self.session.connection(execution_options={"schema_translate_map": {Config.DB_SCHEMA: "tenant_1"}})

  obj_in_data = jsonable_encoder(obj_in)
  db_obj = self.model(**obj_in_data)


  self.session.add(db_obj)

  await self.session.commit();

  await self.session.connection(execution_options={"schema_translate_map": {Config.DB_SCHEMA: "tenant_1"}})

  await self.session.refresh(db_obj)
  return db_obj
```

**Reply by zzzeek** [reply]
when you call `session.commit()`, the connection it uses is returned to the connection pool, so you'd necessarily have to call `.connection()` again like that.     to set execution options per session, we recommend calling this upon the engine that you pass to the session, see https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-isolation-for-individual-sessions

**Reply by zzzeek** [reply]
there may at some point have been a feature request to add execution_options to `Session` constructor itself but i dont know if this ever got tracked

**Reply by Kokleng-Dev** [reply]
Thank you sir, from your informations, i found this.

```
create_async_engine(
    uri,
    connect_args={
        "server_settings": {"search_path": "my_schema"}
    },
)
```

i already test it and it works 🫡.

but i have one question, if i have multi schema (multi-tenant app with difference schema), so i need to create more create_async_engine for difference schema?
is it a good idea?

**Reply by CaselIT** [reply]
replied to it here https://github.com/sqlalchemy/sqlalchemy/discussions/6898#discussioncomment-12644612
